sources/OpenMcdf.Extensions/OLEProperties/PropertyFactory.cs

```diff
@@ -412,26 +412,86 @@ public override string ReadScalarValue(System.IO.BinaryReader br)
                 uint size = br.ReadUInt32();
                 data = br.ReadBytes((int)size);
 
-                return Encoding.GetEncoding(codePage).GetString(data);
+                var result = Encoding.GetEncoding(codePage).GetString(data);
+                //result = result.Trim(new char[] { '\0' });
+
+                //if (this.codePage == CodePages.CP_WINUNICODE)
+                //{
+                //    result = result.Substring(0, result.Length - 2);
+                //}
+                //else
+                //{
+                //    result = result.Substring(0, result.Length - 1);
+                //}
+
+                return result;
             }
 
             public override void WriteScalarValue(BinaryWriter bw, string pValue)
             {
-                data = Encoding.GetEncoding(codePage).GetBytes(pValue);
-                uint dataLength = (uint)data.Length;
+                //bool addNullTerminator = true;
+
+                if (String.IsNullOrEmpty(pValue)) //|| String.IsNullOrEmpty(pValue.Trim(new char[] { '\0' })))
+                {
+                    bw.Write((uint)0);
+                }
+                else if (this.codePage == CodePages.CP_WINUNICODE)
+                {
 
-                // The string data must be null terminated, so add a null byte if there isn't one already
-                bool addNullTerminator =
-                    dataLength == 0 || data[dataLength - 1] != '\0';
+                    data = Encoding.GetEncoding(codePage).GetBytes(pValue);
+
+                    //if (data.Length >= 2 && data[data.Length - 2] == '\0' && data[data.Length - 1] == '\0')
+                    //    addNullTerminator = false;
+
+                    uint dataLength = (uint)data.Length;
+
+                    //if (addNullTerminator)
+                    dataLength += 2;            // null terminator \u+0000

```

Comment:
Possibly the write should be ```bw.Write((uint)dataLength / 2);``` as on line 533? (or possibly it could share the writing code rather than duplicating it?)

Reply:
> MSOLEPS says this should be the length in characters (not bytes) including the null terminator but not including the padding (if any)

No, MS-OLEPS for CodeString says this:

```
Size (4 bytes): The size in bytes of the Characters field, including the null terminator, but not 
including padding (if any). If the property set's CodePage property has the value CP_WINUNICODE 
(0x04B0), then the value MUST be a multiple of 2.
```

The semantic of Size field is different between CodePage String with CP_WINUNICODE and Unicode String